compiler/qsc_parse/src/lex/formatter.rs

```diff
@@ -0,0 +1,270 @@
+
+    // There should always be at least one newline
+    if count_newlines < 1 {
+        count_newlines = 1;
+    }
+    let mut new = "\n".repeat(count_newlines);
+    new.push_str(&make_indent_string(indent_level));
+    new
+}
+
+fn apply_rules(
+    left: &ConcreteToken,
+    whitespace: &str,
+    right: &ConcreteToken,
+    code: &str,
+    indent_level: usize,
+) -> Vec<Edit> {
+    let mut edits = vec![];
+    // when we get here, neither left nor right should be whitespace
+
+    // if the right is a close brace, the indent level should be one less
+    let indent_level = if let ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace)) = right.kind
+    {
+        if indent_level > 0 {
+            indent_level - 1
+        } else {
+            indent_level
+        }
+    } else {
+        indent_level
+    };
+
+    match (&left.kind, &right.kind) {
+        (
+            ConcreteTokenKind::Cooked(TokenKind::Open(l)),
+            ConcreteTokenKind::Cooked(TokenKind::Close(r)),
+        ) if l == r => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (ConcreteTokenKind::Comment | ConcreteTokenKind::Cooked(TokenKind::DocComment), _) => {
+            rule_trim_comments(left, &mut edits, code);
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (ConcreteTokenKind::Cooked(TokenKind::Semi), _) => match &right.kind {
+            ConcreteTokenKind::Comment => {
+                if whitespace.contains('\n') {
+                    rule_indentation(left, whitespace, right, &mut edits, indent_level);
+                }
+            }
+            _ => {
+                rule_indentation(left, whitespace, right, &mut edits, indent_level);
+            }
+        },
+        (_, ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace))) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (
+            _,
+            ConcreteTokenKind::Cooked(TokenKind::Keyword(
+                Keyword::Operation | Keyword::Function | Keyword::Newtype | Keyword::Namespace,
+            )),
+        ) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (ConcreteTokenKind::Cooked(TokenKind::Open(Delim::Brace)), _) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (
+            ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace)),
+            ConcreteTokenKind::Cooked(TokenKind::Semi),
+        ) => {
+            rule_no_space(left, whitespace, right, &mut edits);
+        }
+        (ConcreteTokenKind::Cooked(TokenKind::Close(Delim::Brace)), _) => {
+            rule_indentation(left, whitespace, right, &mut edits, indent_level);
+        }
+        (ConcreteTokenKind::Cooked(cooked_left), ConcreteTokenKind::Cooked(cooked_right)) => {
+            match (cooked_left, cooked_right) {
```

Comment:
Similar to my last comment. Instead of having nested `match`es I think it would be better to write this out line by line, like:

```rs
// abc abc
(CTK::Cooked(TokenKind::Ident), CTK::Cooked(TokenKind::Ident)) => 
(CTK::Cooked(TokenKind::Keyword(_), CTK::Cooked(TokenKind::Ident)) => 
// abc :
(CTK::Cooked(_), CTK::Cooked(TokenKind::Colon)) => 
// : abc
(CTK::Cooked(TokenKind::Colon), CTK::Cooked(_)) => 
// , abc
(CTK::Cooked(TokenKind::Comma), CTK::Cooked(_)) => 
``
 etc.

(You can take it or leave it on the comment suggestion, but I thought it would be helpful to have a visual example of what the rule is)

Reply:
I tried it out, and I think the verbosity of having `CTK::Cooked(...)` everywhere really hurts my ability to quickly get at the relevant information of what two token kinds are being compared.
![image](https://github.com/microsoft/qsharp/assets/55811729/15099364-3a3c-47a6-9787-e2a23ed7e692)
versus
![image](https://github.com/microsoft/qsharp/assets/55811729/18c8df11-8292-456f-827c-2e3bab844110)

Reply:
I like that suggestion a lot, and didn't know Rust supported that. I _did_ have to rename the `TokenKind::String` case though to avoid it conflicting with the built-in `String` type.